fdbserver/storageserver.actor.cpp

```diff
@@ -3521,56 +3521,85 @@ bool rangeQuery(const std::string& s) {
+void preprocessMappedKey(Tuple& mappedKeyFormatTuple,
+                         std::vector<Optional<Tuple>>& vt,
+                         std::vector<std::string>& strings,
+                         bool& isRangeQuery) {
+	vt.reserve(mappedKeyFormatTuple.size());
+	strings.reserve(mappedKeyFormatTuple.size());
 
-	Tuple mappedKeyTuple;
 	for (int i = 0; i < mappedKeyFormatTuple.size(); i++) {
 		Tuple::ElementType type = mappedKeyFormatTuple.getType(i);
 		if (type == Tuple::BYTES || type == Tuple::UTF8) {
 			std::string s = mappedKeyFormatTuple.getString(i).toString();
 			auto sz = s.size();
 			bool escaped = unescapeLiterals(s, "{{", "{");
 			escaped = unescapeLiterals(s, "}}", "}") || escaped;
+			strings.push_back(s);
 			if (escaped) {
-				mappedKeyTuple.append(s);
+				Tuple escapedTuple;
+				escapedTuple.append(s);
+				vt.push_back(Optional<Tuple>(escapedTuple));
 			} else if (singleKeyOrValue(s, sz)) {
-				int idx;
-				Tuple* referenceTuple;
-				try {
-					idx = std::stoi(s.substr(3, sz - 5));
-				} catch (std::exception& e) {
-					throw mapper_bad_index();
-				}
-				if (s[1] == 'K') {
-					unpackKeyTuple(&referenceTuple, keyTuple, keyValue);
-				} else if (s[1] == 'V') {
-					unpackValueTuple(&referenceTuple, valueTuple, keyValue);
-				} else {
-					ASSERT(false);
-					throw internal_error();
-				}
-				if (idx < 0 || idx >= referenceTuple->size()) {
-					throw mapper_bad_index();
-				}
-				mappedKeyTuple.append(referenceTuple->subTuple(idx, idx + 1));
+				// when it is SingleKeyOrValue, insert an empty Tuple to vector as placeholder
+				vt.push_back(Optional<Tuple>(Tuple()));
 			} else if (rangeQuery(s)) {
-				if (i != mappedKeyFormatTuple.size() - 1) {
-					// It must be the last element of the mapper tuple
-					throw mapper_bad_range_decriptor();
-				}
-				// Every record will try to set it. It's ugly, but not wrong.
+				// when it is rangeQuery, insert Optional.empty as placeholder
+				vt.push_back(Optional<Tuple>());
 				isRangeQuery = true;
-				// Do not add it to the mapped key.
 			} else {
-				// If the element is a string but neither escaped nor descriptors, add to result.
-				mappedKeyTuple.append(mappedKeyFormatTuple.subTuple(i, i + 1));
+				vt.push_back(Optional<Tuple>(mappedKeyFormatTuple.subTuple(i, i + 1)));
 			}
 		} else {
-			// If the element not a string, add to result.
-			mappedKeyTuple.append(mappedKeyFormatTuple.subTuple(i, i + 1));
+			vt.push_back(Optional<Tuple>(mappedKeyFormatTuple.subTuple(i, i + 1)));
+		}
+	}
+}
+
+Key constructMappedKey(KeyValueRef* keyValue,
+                       std::vector<Optional<Tuple>>& vec,
+                       Tuple& mappedKeyTuple,
+                       std::vector<std::string>& strings) {
+	// Lazily parse key and/or value to tuple because they may not need to be a tuple if not used.
+	Optional<Tuple> keyTuple;
+	Optional<Tuple> valueTuple;
+	mappedKeyTuple.clear();
+	mappedKeyTuple.reserve(vec.size());
```

Comment:
adding this as std::vector::clear might reallocate the space https://www.cplusplus.com/reference/vector/vector/clear/ `A reallocation is not guaranteed to happen, and the [vector capacity](https://www.cplusplus.com/vector::capacity) is not guaranteed to change due to calling this function.`  @sears

Reply:
Thanks.  That's an annoying nuance of the API.  (I can think of no reason to use vector::clear, other than to avoid allocator load.)